tests/infer/reparam/test_conjugate.py

```diff
@@ -0,0 +1,210 @@
+    assert data.shape == batch_shape + (num_steps, obs_dim)
+    prior = dist.GaussianHMM(init_dist, trans_mat, trans_dist, obs_mat, obs_dist)
+    likelihood = dist.Normal(data, 1).to_event(2)
+    posterior, log_normalizer = prior.conjugate_update(likelihood)
+
+    def model(data):
+        with pyro.plate_stack("plates", batch_shape):
+            z = pyro.sample("z", prior)
+            pyro.sample("x", dist.Normal(z, 1).to_event(2), obs=data)
+
+    def guide(data):
+        with pyro.plate_stack("plates", batch_shape):
+            pyro.sample("z", posterior)
+
+    reparam_model = poutine.reparam(model, {"z": ConjugateReparam(likelihood)})
+
+    def reparam_guide(data):
+        pass
+
+    elbo = Trace_ELBO(num_particles=1000, vectorize_particles=True)
+    expected_loss = elbo.differentiable_loss(model, guide, data)
+    actual_loss = elbo.differentiable_loss(reparam_model, reparam_guide, data)
+    assert_close(actual_loss, expected_loss, atol=0.01)
+
+    params = [trans_mat, obs_mat]
+    expected_grads = torch.autograd.grad(expected_loss, params, retain_graph=True)
+    actual_grads = torch.autograd.grad(actual_loss, params, retain_graph=True)
+    for a, e in zip(actual_grads, expected_grads):
+        assert_close(a, e, rtol=0.01)
+
+
+def random_stable(shape):
+    stability = dist.Uniform(1.4, 1.9).sample(shape)
+    skew = dist.Uniform(-1, 1).sample(shape)
+    scale = torch.rand(shape).exp()
+    loc = torch.randn(shape)
+    return dist.Stable(stability, skew, scale, loc)
+
+
+@pytest.mark.parametrize("batch_shape", [(), (4,), (3, 2)], ids=str)
+@pytest.mark.parametrize("hidden_dim,obs_dim", [(1, 1), (2, 3)], ids=str)
+@pytest.mark.parametrize("num_steps", range(1, 6))
+def test_stable_hmm_smoke(batch_shape, num_steps, hidden_dim, obs_dim):
+    init_dist = random_stable(batch_shape + (hidden_dim,)).to_event(1)
+    trans_mat = torch.randn(batch_shape + (num_steps, hidden_dim, hidden_dim), requires_grad=True)
+    trans_dist = random_stable(batch_shape + (num_steps, hidden_dim)).to_event(1)
+    obs_mat = torch.randn(batch_shape + (num_steps, hidden_dim, obs_dim), requires_grad=True)
+    obs_dist = random_stable(batch_shape + (num_steps, obs_dim)).to_event(1)
+    data = obs_dist.sample()
+    assert data.shape == batch_shape + (num_steps, obs_dim)
+
+    def model(data):
+        hmm = dist.LinearHMM(init_dist, trans_mat, trans_dist, obs_mat, obs_dist)
+        with pyro.plate_stack("plates", batch_shape):
+            z = pyro.sample("z", hmm)
+            pyro.sample("x", dist.Normal(z, 1).to_event(2), obs=data)
+
+    # Test that we can combine these two reparameterizers.
+    reparam_model = poutine.reparam(model, {
+        "z": LinearHMMReparam(StableReparam(), StableReparam(), StableReparam()),
+    })
+    reparam_model = poutine.reparam(reparam_model, {
+        "z": ConjugateReparam(dist.Normal(data, 1).to_event(2)),
+    })
+    reparam_guide = AutoDiagonalNormal(reparam_model)  # Models auxiliary variables.
```

Comment:
This test is the culmination of months of reparameterizer work. It starts with an inhomogeneous Levy Stable process, applies `LinearHMMReparam` which calls `StableReparam` three times under the hood resulting in a conditional `GaussianHMM`, then applies `ConjugateReparam` to make a conjugate update to the conditional `GaussianHMM`, and finally uses an autoguide to model the auxiliary variables introduced by `StableReparam()`. Running the model calls parallel-scan forward-filter backward-sample. Hurray for composable inference 🎉